controllers/jenkins/pipelinerun/pipelinerun_controller.go

```diff
@@ -166,47 +166,82 @@ func (r *Reconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
+		return ctrl.Result{}, err
+	} else if hasPending {
+		return ctrl.Result{}, nil
+	}
+
 	// set Jenkins run ID
-	if pr.Annotations == nil {
-		pr.Annotations = make(map[string]string)
+	if pipelineRunCopied.Annotations == nil {
+		pipelineRunCopied.Annotations = make(map[string]string)
 	}
-	pr.Annotations[prv1alpha3.JenkinsPipelineRunIDKey] = pipelineBuild.ID
+	pipelineRunCopied.Annotations[prv1alpha3.JenkinsPipelineRunIDKey] = pipelineBuild.ID
 
 	// the Update method only updates fields except subresource: status
-	if err := r.updateLabelsAndAnnotations(ctx, &pr); err != nil {
+	if err := r.updateLabelsAndAnnotations(ctx, pipelineRunCopied); err != nil {
 		log.Error(err, "unable to update PipelineRun labels and annotations.")
 		return ctrl.Result{}, err
 	}
 
-	pr.Status.StartTime = &v1.Time{Time: time.Now()}
-	pr.Status.UpdateTime = &v1.Time{Time: time.Now()}
+	pipelineRunCopied.Status.StartTime = &v1.Time{Time: time.Now()}
+	pipelineRunCopied.Status.UpdateTime = &v1.Time{Time: time.Now()}
 	// due to the status is subresource of PipelineRun, we have to update status separately.
 	// see also: https://book-v1.book.kubebuilder.io/basics/status_subresource.html
 
-	if err := r.updateStatus(ctx, &pr.Status, req.NamespacedName); err != nil {
+	if err := r.updateStatus(ctx, &pipelineRunCopied.Status, req.NamespacedName); err != nil {
 		log.Error(err, "unable to update PipelineRun status.")
 		return ctrl.Result{}, err
 	}
-	r.recorder.Eventf(&pr, corev1.EventTypeNormal, prv1alpha3.Started, "Started PipelineRun %s", req.NamespacedName)
+	r.recorder.Eventf(pipelineRunCopied, corev1.EventTypeNormal, prv1alpha3.Started, "Started PipelineRun %s", req.NamespacedName)
 	// requeue after 1 second
 	return ctrl.Result{RequeueAfter: 1 * time.Second}, nil
 }
 
+// hasSamePendingPipelineRun checks if there is still a PipelineRun with the same run ID.
+func (r *Reconciler) hasSamePendingPipelineRun(runID string, pipelineRunKey client.ObjectKey, log logr.Logger) (bool, error) {
+	// check if the run ID exists in the PipelineRun
+	pipelineRuns := &prv1alpha3.PipelineRunList{}
+	if err := r.Client.List(context.Background(), pipelineRuns,
+		client.HasLabels{prv1alpha3.JenkinsPipelineRunIDKey},
+		client.MatchingLabels{prv1alpha3.JenkinsPipelineRunIDKey: runID}); err != nil {
+		log.Error(err, "unable to search all PipelineRuns with a specific run ID", "runID", runID)
+		return false, err
+	}
+	if len(pipelineRuns.Items) == 0 {
+		return false, nil
+	}
+	// if we find the PipelineRuns with the same run ID,
+	// then we will delete this one
+	pipelineRunToDelete := &prv1alpha3.PipelineRun{}
+	if err := r.Client.Get(context.Background(), pipelineRunKey, pipelineRunToDelete); err != nil {
+		log.Error(err, "uanble to get PipelineRun to be deleted")
+		return false, err
+	}
+	if err := r.Client.Delete(context.Background(), pipelineRunToDelete); err != nil {
```

Comment:
You should not delete anything in function `hasSamePendingPipelineRun`. Because I cannot read it from the function name.

Reply:
Thank you for pointing out this. I've made some changes for this, please take a look at it.